File: Necklace.cpp
/*-*-C++-*-*////////////////////////////////////////////////////////////////////
#include /*                                                        ***/ <cstdio>
#include /*                                                        ***/ <string>
#include /*                                                        */ <iostream>
#include /*                                                        ***/ <vector>
#include /*                                                        */ <iterator>
#include /*                                                        ****/ <tuple>
using namespace                                                             std;
////////////////////////////////////////////////////////////////////////////////
////----____oooo````;;;;,,,,....''''----    ....////---->>>>!!!!^^^^    ____%%%%
string::iterator next(string::iterator it, string &s)
{ /* Implement circular iteration by overloading std::next */
  return ( ++it == s.end() )? s.begin(): it; }
  
string::iterator prev(string::iterator it, string &s)
{ /* oo */
  return ( it-- == s.begin() )? (s.end() - 1): it; /* neva beginze */ }

////    ----    <<<<    ****    ||||    ````    ,,,,    ....    ;;;;    >>>>
vector<int> extract_wpos(const string &s)
{ /* Extract white positions from s. */
  vector<int> wpos;
  int siz = s.size();
  for (int i(0); i < siz; ++i) {
    if( s[i] == 'w' ){ wpos.push_back(i); }}
  return wpos; }

////----____````////====,,,,''''////####....""""////eeee    ||||////&&&&____''''
template <typename T>
ostream &operator<<(ostream &out, const vector<T> &vec)
{ /* Yea! */
  copy(vec.begin(), vec.end(), ostream_iterator<T>(out, " "));
  return out; }

////////////////////////////////////////////////////////////////////////////////
vector<string> paint(const string &s)
{ /* Use w's in s to generate all possible coloured necklaces. */
  vector<int> wpos = extract_wpos(s);
  int wsiz = wpos.size(); // number of w's
  int car  = 1 << wsiz;   // cardinality of the set of all wsiz subsets
  vector<string> stk;     // buffer 
  /* [Ok] let's have a string s = wrbw, than wpos should be [0,3] and
   * car = 4, now we have to match the 1-bits in 0 <= bits < car to red
   * and 0-bits to b in the corresponding w's:
   * wpos: 0 '   '   ' 3 
   * s   : w ' r ' b ' w
   * bits: 0 '   '   ' 0 : 0(brbb)
   *       1 '   '   ' 0 : 1(rrbb)
   *       0 '   '   ' 1 : 2(brbr)
   *       1 '   '   ' 1 : 3(rrbr)
   */
  for (int bits = 0; bits < car; ++bits) {
    string t = s; /* copy */
    for (int j = 0; j < wsiz; ++j) {
      int p = wpos[j];
      if( bits & (1 << j) ){
        t[p] = 'r'; 
      } else {
        t[p] = 'b'; 
      }}
    stk.push_back(t); }
  return stk; }

////////////////////////////////////////////////////////////////////////////////
int cont(const string::iterator &fwd, string &s)
{ /* Count number of same beads on both, fwd and bwd ends. */
  int x = 2; // conter
  const string::iterator &bwd = ::prev(fwd, s); 
  for(string::iterator it = fwd; *(it = ::next(it, s)) == *fwd; ++x)
    ;
  for(string::iterator it = bwd; *(it = ::prev(it, s)) == *bwd; ++x)
    ;
  return x; }

    ////    ----    ````    ,,,,    ____    ====    ||||    """"    ****    ;;;;
pair<int, int> junction(string &s)
{ /* Apply cont to all rb or br junctions, return maximum cont, and position. */
  int mx = 0; // maximum count
  int m = -1; // mx position
  for (auto it = s.begin(); it != s.end(); ++it) {
    if (*it != *::prev(it, s)) { // junction
      int x = cont(it, s);
      if( x > mx ){
        m = it - s.begin();
        mx = x; }}}
  if( m == -1 ){ // there are no junctions
    mx = s.size(); }
  return {mx, m}; }

////    ----****____\\\\----    ****____  * ____ *  ==== *  ````*   ====  * ====
tuple<int, int, string> Necklace(string &s)
{ /* Return beads counter, cut position and necklace string. */
  vector<string> stk = paint(s);
  int cont(0);
  int j;
  vector<string>::iterator p;
  for (auto it = stk.begin(); it != stk.end(); ++it) {
    auto [mx, m] = junction(*it);
    if( cont < mx ){
      cont = mx;
      j = m;
      p = it; }}
  return {cont, j, *p}; }

////////////////////////////////////////////////////////////////////////////////
int main(int argc, char *argv[])
{ /* yea! */
  string s = "rwbwwbbbrrwwbrr";
  puts(s.c_str());
  auto [cont, j, Nec] = Necklace(s);
  cout << cont << ' ' << j << ' ' << Nec; 
  puts(""); }

/////////////////////////////////////////////////////////////////////////// log:
// xògu neE

File: Necklace.py
#!/usr/bin/env python3
####----\\\\;;;;    ____````,,,,||||====>>>>;;;;....////
def getColours(patron, nofBeads):
  """ patron is an integer with nofBeads number of bits.
      In binary it represents  a colouring pattern, e.g:
      patron = 10, nofBead = 4 is 1010. Now 1 is 'r' and
      0 is 'b', zo ..  **   the function will return the 
      corresponding bytearray, in the above case 'brbr'.
      Note that it's reversed.  oo     --  ..   ^^    """
  getColours.clr = b'br'
  b = bytearray(nofBeads)
  for pos in range(nofBeads):
    bit = ((1 << pos) & patron) >> pos
    b[pos] = getColours.clr[bit]
  return b

####====\\\\::::    ****____^^^^----,,,,>>>>    %%%%____
def findWhiteBeads(s):
  """ Return the positions of white beads. """
  ls = []
  for pos, char in enumerate( s ):
    if char == 'w': ls.append( pos )
  return ls

####....````____====''''----____====    ^^^^----____%%%%
def paint(wPos, t, patron, nofBeads):
  """ wPos is a list of w positions of t, which is a
      bytearray copy of the orininal template necklace
      string.                                        """
  b = getColours(patron, nofBeads)
  for j in range( nofBeads ):
    t[wPos[j]] = b[j]
    
####____    ====````----^^^^____====>>>>    ''''____````
####====oooo>>>><<<<====    ::::    ----^^^^____====>>>>
####----////\\\\....````====____****____====''''....,,,,
####====    ====____%%%%....====____====''''"""";;;;____
####    ****====,,,,____!!!!;;;;    ````____====****
####....    ''''    ____====;;;;""""____    ====----////
####....,,,,____====\\\\////____^^^^____====____^^^^----
####____````****oooo____////----%%%%####~~~~xxxx____****
####    ====****----,,,,%%%%        @@@@====____||||
####----$$$$....''''////----4444    ----____""""''''////
def clcknext(j, n):
  """ solid as a [s]ock """
  j += 1
  if j == n: j = 0
  return j

def clckprev(j, n):
  """ bo[+]m """
  j -= 1
  # w=[ 1, 2, 3 ], w[-3] iz 1, but w[-4] is boom. 
  if j < -n: j = -1
  return j

####====    ____::::""""''''````^^^^||||____====    ////
# s - a string: "bbbrrrb" of blue and red beads.     _=
# j - junction posi ^ tion                         _/
# retour number of same beads on both ends      -//
def cont(s, j):
  """ x [o] g u   n e E """
  # at least we have 2:
  x = 2
  bwd = j - 1 # backward position
  fwd = j     # w h a t i s t h i s ?
  n = len(s)
  while True:
    prev = clckprev(bwd, n)
    if s[prev] != s[bwd]: break
    x += 1
    bwd = prev
  while True:
    next = clcknext(fwd, n)
    if s[next] != s[fwd]: break
    x += 1
    fwd = next  
  return x

####____!!!!    \\\\^^^^////____````!!!!>>>>____'''',,,,
def getJunctions(s):
  """ 6 e 3   g y M u """
  junc = []
  for j in range(len(s)):
    if s[j] != s[j - 1]:
      junc.append(j)
  return junc 

####====>>>>====<<<<====````====,,,,====((((====''''====
def maxCont(s):
  """          """
  mx = 0  # max cont
  mj = -1 # mx junction 
  for j in getJunctions(s):
    x = cont(s, j)
    if mx < x:
      mx, mj = x, j
  if mj < 0: mx = len(s)
  return mx, mj

if __name__ == '__main__':
  s = 'wrwbwr'
  print(s)
  wpos = findWhiteBeads(s)
  nofw = len(wpos)
  mx, mj, mt = 0, None, None
  for j in range( 1 << nofw ):
    t = bytearray(s, encoding='utf-8')
    paint(wpos, t, j, nofw)
    x, j = maxCont(t.decode())
    if mx < x: mx, mj, mt = x, j, t
  print(mx, mj, mt.decode())

####____''''====>>>>"""",,,,    !!!!````~~~~%%%%,,,,log:
